### PR TITLE
Trie Proof Nodes and Iterate

### PR DESCRIPTION
**Pull request by asdacap**
- Basic iterate leaf, generate and verify proofs. 
- Iterate using trie visit. Dont wanna mess around with multiple implementation of storage right now.
- Needed for snap sync.

### CODE REVIEW

File: core/trie/trie.go [hunk 1]
```diff
@@ -192,16 +192,143 @@ func (t *Trie) Get(key *felt.Felt) (*felt.Felt, error) {
...
+		}
+	}
+
+	return t.doIterate(startValue, thenode.Right, consumer)
+}
+
+type ProofNode struct {
+	Key  *Key
+	Hash *felt.Felt
+}
```
Comment (omerfirmak) [inline core/trie/trie.go:242]: We need to come up with a standard way of representing a proof node across different implementation.
I am not sure if pathfinder will be able to consume this proof.

We probably need to use the definition from the spec. https://docs.starknet.io/documentation/architecture_and_concepts/State/starknet-state/#specifications

What spec defines as `path` is a bit different. We calculate that in `path(key, parentKey *Key) Key`

Comment (omerfirmak) [reply core/trie/trie.go:242]: https://github.com/starknet-io/starknet-p2p-specs/pull/8/files#diff-1895f60c4bbe4eec02ba055cd529fd9ad7cb4b76bad983cbc68a2d1dc0ce0190R6-R21

Comment (asdacap) [reply core/trie/trie.go:242]: The spec specification for path seems to be for calculating the commitment, not the actual tree path. You can assume this as `JunoSnapProofNode` if you want and rename it later. As for the p2p spec, I thought we would rather make it juno specific first. Or would you rather have me dump the whole snap prototype branch and you guys can figureout how to implement it?

Comment (asdacap) [reply core/trie/trie.go:242]: I'll assume you'll adapt it accordingly.

**Pull request closed without merging**